keras_nlp/tokenizers/tokenizer.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class Tokenizer(keras.layers.Layer):
+    """A base class for tokenizer layers.
+
+    The class is intended as a base class when implementing a tokenizer as a
+    `keras.layers.Layer`. It contains two new methods `tokenize` and
+    `detokenize`.
+
+    Subclassers should always implement the `tokenize` method, which will also
+    also be the default when invoking the layer directly on inputs
+
+    If a layer does not support detokenization (the tokenization step is not
+    reversible), the `detokenize` method can be skipped.
+
+    Examples:
+
+    ```python
+    class WhitespaceTokenizer(keras_nlp.tokenizers.Tokenizer):
+        def tokenize(self, inputs: tf.Tensor) -> tf.Tensor:
+            return tf.strings.split(inputs).to_tensor()
+
+        def detokenizer(self, inputs: tf.Tensor) -> tf.Tensor:
+            return tf.strings.reduce_join([inputs], separator=" ", axis=-1)
+
+    tokenizer = WhitespaceTokenizer()
+    ```
+    """
+
+    def __new__(cls, *args, **kwargs):
+        # Wrap the `tokenize` and `detokenize` methods so they route through
+        # __call__. This is needed for functional model support.
+        obj = super().__new__(cls, *args, **kwargs)
+        obj._tokenize_without_call = obj.tokenize
+        obj._detokenize_without_call = obj.detokenize
+        obj.tokenize = obj._tokenize_with_call
+        obj.detokenize = obj._detokenize_with_call
+        return obj
+
+    def tokenize(self, inputs: tf.Tensor, *args, **kwargs) -> tf.Tensor:
+        """Transform input tensors of strings into output tokens.
+
+        Args:
+            inputs: Input tensor, or dict/list/tuple of input tensors.
+            *args: Additional positional arguments.
+            **kwargs: Additional keyword arguments.
+        """
+        raise NotImplementedError(
+            "No implementation of `tokenize` was found for "
```

Comment:
Nit: refer to methods/functions with a `()` at the end to make it clear they're methods/functions

Reply:
Done